File: packages/ios/Sources/CivUI/CivYesNo.swift
// CivUI — CivYesNo for SwiftUI
// Accessible yes/no binary choice following government design system patterns.
// Renders: legend → hint → error → button pair (Section 508 compliant)

import SwiftUI

/// Accessible yes/no binary choice for government applications.
///
/// Renders two styled buttons (Yes/No) in an HStack. The selected button
/// uses a primary background with white text; the unselected button uses
/// a white background with a border. Mirrors the web `civ-yes-no` component.
///
/// VoiceOver announces "Yes, selected" or "No, selected" for the active choice.
///
/// Usage:
/// ```swift
/// CivYesNo(
///     legend: "Are you a U.S. citizen?",
///     value: $isCitizen,
///     hint: "Select one option"
/// )
/// ```
public struct CivYesNo: View {
    // MARK: - Properties

    /// Legend text displayed above the button pair.
    public let legend: String

    /// Bound value: "yes", "no", or "" (no selection).
    @Binding public var value: String

    /// Help text shown below the legend.
    public var hint: String?

    /// Error message. When set, renders with VoiceOver announcement.
    public var error: String?

    /// Whether the field is required.
    public var isRequired: Bool

    /// Whether the field is disabled.
    public var isDisabled: Bool

    /// Whether the field is read-only (visible but non-interactive).
    public var isReadonly: Bool

    /// Custom label for the "Yes" button.
    public var yesLabel: String

    /// Custom label for the "No" button.
    public var noLabel: String

    /// Optional label for a third button. When non-empty, a third option is rendered.
    public var unsureLabel: String

    /// Form value for the third option (default: "unsure").
    public var unsureValue: String

    /// Called on value change (parallels `civ-change` event).
    public var onChange: ((String) -> Void)?

    /// Called for analytics tracking (parallels `civ-analytics` event).
    public var onAnalytics: ((String, [String: Any]?) -> Void)?

    /// Optional form state for centralized validation.
    public var formState: CivFormState?

    /// Field name for form state registration.
    public var formName: String?

    /// Custom required message for form validation.
    public var requiredMessage: String?

    /// Custom validation function. Returns error string or nil.
    public var formValidate: (() -> String?)?

    /// Whether the field contains PII (excluded from getFormData).
    public var isPii: Bool

    // MARK: - Internal State

    @Environment(\.colorScheme) private var colorScheme

    // MARK: - Initializer

    public init(
        legend: String,
        value: Binding<String>,
        hint: String? = nil,
        error: String? = nil,
        isRequired: Bool = false,
        isDisabled: Bool = false,
        isReadonly: Bool = false,
        yesLabel: String = "Yes",
        noLabel: String = "No",
        unsureLabel: String = "",
        unsureValue: String = "unsure",
        onChange: ((String) -> Void)? = nil,
        onAnalytics: ((String, [String: Any]?) -> Void)? = nil,
        formState: CivFormState? = nil,
        formName: String? = nil,
        requiredMessage: String? = nil,
        formValidate: (() -> String?)? = nil,
        isPii: Bool = false
    ) {
        self.legend = legend
        self._value = value
        self.hint = hint
        self.error = error
        self.isRequired = isRequired
        self.isDisabled = isDisabled
        self.isReadonly = isReadonly
        self.yesLabel = yesLabel
        self.noLabel = noLabel
        self.unsureLabel = unsureLabel
        self.unsureValue = unsureValue
        self.onChange = onChange
        self.onAnalytics = onAnalytics
        self.formState = formState
        self.formName = formName
        self.requiredMessage = requiredMessage
        self.formValidate = formValidate
        self.isPii = isPii
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: CivTokens.Spacing._1) {
            // 1. Legend
            legendView

            // 2. Hint
            if let hint, !hint.isEmpty {
                Text(hint)
                    .font(.system(size: CivTokens.Typography.FontSize.sm))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Base.dark,
                        dark: CivTokens.DarkColors.Base.dark
                    ))
                    .accessibilityIdentifier("civ-hint")
            }

            // 3. Error
            if let error, !error.isEmpty {
                Text(error)
                    .font(.system(size: CivTokens.Typography.FontSize.sm,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityIdentifier("civ-error")
            }

            // 4. Button pair
            HStack(spacing: CivTokens.Spacing._3) {
                choiceButton(label: yesLabel, choiceValue: "yes")
                choiceButton(label: noLabel, choiceValue: "no")
                if !unsureLabel.isEmpty {
                    choiceButton(label: unsureLabel, choiceValue: unsureValue)
                }
            }
        }
        .padding(.bottom, CivTokens.Spacing._4)
        .disabled(isDisabled || isReadonly)
        .opacity(isDisabled ? 0.5 : 1.0)
        .accessibilityElement(children: .contain)
        .onChange(of: error) { newError in
            if let newError, !newError.isEmpty {
                UIAccessibility.post(notification: .announcement, argument: newError)
            }
        }
        .onAppear { registerWithFormState() }
        .onDisappear { unregisterFromFormState() }
    }

    // MARK: - Subviews

    private var legendView: some View {
        HStack(spacing: CivTokens.Spacing._0_5) {
            Text(legend)
                .font(.system(size: CivTokens.Typography.FontSize.lg,
                              weight: CivTokens.Typography.FontWeight.bold))
                .foregroundColor(adaptiveColor(
                    light: CivTokens.Colors.Base.darkest,
                    dark: CivTokens.DarkColors.Base.darkest
                ))

            if isRequired {
                Text("*")
                    .font(.system(size: CivTokens.Typography.FontSize.lg,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityLabel("required")
            }
        }
    }

    private func choiceButton(label: String, choiceValue: String) -> some View {
        let isSelected = value == choiceValue

        return Button(action: {
            guard !isDisabled && !isReadonly else { return }
            value = choiceValue
            onChange?(choiceValue)
            onAnalytics?("change", ["value": choiceValue])
        }) {
            Text(label)
                .font(.system(size: CivTokens.Typography.FontSize.base,
                              weight: CivTokens.Typography.FontWeight.semibold))
                .padding(.horizontal, CivTokens.Spacing._6)
                .padding(.vertical, CivTokens.Spacing._2_5)
                .foregroundColor(isSelected
                    ? adaptiveColor(light: CivTokens.Colors.White.default_,
                                    dark: CivTokens.DarkColors.White.default_)
                    : adaptiveColor(light: CivTokens.Colors.Primary.default_,
                                    dark: CivTokens.DarkColors.Primary.default_))
                .background(isSelected
                    ? adaptiveColor(light: CivTokens.Colors.Primary.default_,
                                    dark: CivTokens.DarkColors.Primary.default_)
                    : adaptiveColor(light: CivTokens.Colors.White.default_,
                                    dark: CivTokens.DarkColors.White.default_))
                .cornerRadius(CivTokens.Border.Radius.default_)
                .overlay(
                    RoundedRectangle(cornerRadius: CivTokens.Border.Radius.default_)
                        .stroke(
                            adaptiveColor(
                                light: CivTokens.Colors.Primary.default_,
                                dark: CivTokens.DarkColors.Primary.default_
                            ),
                            lineWidth: CivTokens.Border.Width._2
                        )
                )
        }
        .accessibilityLabel("\(label)\(isSelected ? ", selected" : "")")
        .accessibilityAddTraits(isSelected ? [.isSelected] : [])
    }

    // MARK: - Form State Registration

    private func registerWithFormState() {
        guard let formState, let formName, !formName.isEmpty else { return }
        formState.register(CivFormState.CivFieldRegistration(
            name: formName,
            label: legend,
            getValue: { value },
            setValue: { newValue in value = newValue },
            isRequired: isRequired,
            requiredMessage: requiredMessage ?? "",
            validate: formValidate,
            isPii: isPii
        ))
    }

    private func unregisterFromFormState() {
        guard let formState, let formName, !formName.isEmpty else { return }
        formState.unregister(formName)
    }

    // MARK: - Color Helper

    private func adaptiveColor(light: Color, dark: Color) -> Color {
        colorScheme == .dark ? dark : light
    }
}

// MARK: - Preview

#if DEBUG
struct CivYesNo_Previews: PreviewProvider {
    struct PreviewWrapper: View {
        @State private var citizen = ""
        @State private var veteran = "yes"

        var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    CivYesNo(
                        legend: "Are you a U.S. citizen?",
                        value: $citizen,
                        hint: "Select one option",
                        isRequired: true
                    )

                    CivYesNo(
                        legend: "Are you a veteran?",
                        value: $veteran
                    )

                    CivYesNo(
                        legend: "Do you have a service-connected disability?",
                        value: $citizen,
                        unsureLabel: "I'm not sure"
                    )

                    CivYesNo(
                        legend: "Disabled question",
                        value: .constant(""),
                        isDisabled: true
                    )

                    CivYesNo(
                        legend: "With error",
                        value: .constant(""),
                        error: "Please select an option",
                        isRequired: true
                    )
                }
                .padding()
            }
        }
    }

    static var previews: some View {
        PreviewWrapper()
            .previewDisplayName("Light")
        PreviewWrapper()
            .preferredColorScheme(.dark)
            .previewDisplayName("Dark")
    }
}
#endif

File: packages/android/src/main/kotlin/gov/civui/components/CivYesNo.kt
// CivUI — CivYesNo for Jetpack Compose
// Accessible yes/no binary choice with two styled buttons.
// Renders: legend -> hint -> error -> button pair (Section 508 compliant)

package gov.civui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import gov.civui.tokens.CivTokens

/**
 * Accessible yes/no binary choice for government applications.
 *
 * Renders two buttons in a Row. The selected button uses a filled primary
 * style; the unselected button uses an outlined style. Mirrors the web
 * `civ-yes-no` component.
 *
 * TalkBack announces "Yes, selected" or "No, selected" for the active choice.
 *
 * Usage:
 * ```kotlin
 * var isCitizen by remember { mutableStateOf("") }
 * CivYesNo(
 *     legend = "Are you a U.S. citizen?",
 *     value = isCitizen,
 *     onValueChange = { isCitizen = it },
 *     hint = "Select one option",
 * )
 * ```
 */
@Composable
fun CivYesNo(
    legend: String,
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    hint: String = "",
    error: String = "",
    required: Boolean = false,
    disabled: Boolean = false,
    readonly: Boolean = false,
    yesLabel: String = "Yes",
    noLabel: String = "No",
    unsureLabel: String = "",
    unsureValue: String = "unsure",
    skipLabel: String = "",
    skipValue: String = "",
    onChange: ((String) -> Unit)? = null,
    onAnalytics: ((event: String, data: Map<String, Any>?) -> Unit)? = null,
    name: String = "",
    formState: CivFormState? = null,
) {
    val isDark = isSystemInDarkTheme()

    // Form state registration
    var formError by remember { mutableStateOf("") }
    val effectiveError = error.ifEmpty { formError }

    if (formState != null && name.isNotEmpty()) {
        androidx.compose.runtime.DisposableEffect(name) {
            formState.register(CivFormState.CivFieldRegistration(
                name = name,
                getValue = { value },
                setValue = { onValueChange(it) },
                isRequired = required,
                getError = { formError },
                setError = { formError = it },
            ))
            onDispose { formState.unregister(name) }
        }
    }

    val labelColor = if (isDark) CivTokens.DarkColors.Base.darkest else CivTokens.Colors.Base.darkest
    val hintColor = if (isDark) CivTokens.DarkColors.Base.dark else CivTokens.Colors.Base.dark
    val errorColor = if (isDark) CivTokens.DarkColors.Error.default_ else CivTokens.Colors.Error.default_
    val primaryColor = if (isDark) CivTokens.DarkColors.Primary.default_ else CivTokens.Colors.Primary.default_
    val whiteColor = if (isDark) CivTokens.DarkColors.White.default_ else CivTokens.Colors.White.default_

    Column(
        modifier = modifier
            .padding(bottom = CivTokens.Spacing._4)
            .alpha(if (disabled) 0.5f else 1f)
            .semantics {
                contentDescription = buildString {
                    append(legend)
                    if (required) append(", required")
                    if (effectiveError.isNotEmpty()) append(". Error: $effectiveError")
                }
                if (effectiveError.isNotEmpty()) {
                    error(effectiveError)
                }
            },
    ) {
        // 1. Legend
        CivLabel(
            label = legend,
            required = required,
            labelColor = labelColor,
            errorColor = errorColor,
        )

        // 2. Hint
        CivHint(text = hint.ifEmpty { null }, color = hintColor)

        // 3. Error
        CivError(text = effectiveError.ifEmpty { null }, color = errorColor)

        // 4. Button pair
        Row(
            horizontalArrangement = Arrangement.spacedBy(CivTokens.Spacing._3),
            modifier = Modifier.padding(top = CivTokens.Spacing._1),
        ) {
            // Yes button
            ChoiceButton(
                label = yesLabel,
                isSelected = value == "yes",
                primaryColor = primaryColor,
                whiteColor = whiteColor,
                enabled = !disabled && !readonly,
                onClick = {
                    onValueChange("yes")
                    onChange?.invoke("yes")
                    onAnalytics?.invoke("change", mapOf("value" to "yes"))
                },
            )

            // No button
            ChoiceButton(
                label = noLabel,
                isSelected = value == "no",
                primaryColor = primaryColor,
                whiteColor = whiteColor,
                enabled = !disabled && !readonly,
                onClick = {
                    onValueChange("no")
                    onChange?.invoke("no")
                    onAnalytics?.invoke("change", mapOf("value" to "no"))
                },
            )

            // Optional unsure button
            if (unsureLabel.isNotEmpty()) {
                ChoiceButton(
                    label = unsureLabel,
                    isSelected = value == unsureValue,
                    primaryColor = primaryColor,
                    whiteColor = whiteColor,
                    enabled = !disabled && !readonly,
                    onClick = {
                        onValueChange(unsureValue)
                        onChange?.invoke(unsureValue)
                        onAnalytics?.invoke("change", mapOf("value" to unsureValue))
                    },
                )
            }
        }
    }
}

@Composable
private fun ChoiceButton(
    label: String,
    isSelected: Boolean,
    primaryColor: androidx.compose.ui.graphics.Color,
    whiteColor: androidx.compose.ui.graphics.Color,
    enabled: Boolean,
    onClick: () -> Unit,
) {
    if (isSelected) {
        Button(
            onClick = onClick,
            enabled = enabled,
            colors = ButtonDefaults.buttonColors(
                containerColor = primaryColor,
                contentColor = whiteColor,
            ),
            modifier = Modifier.semantics {
                selected = true
                role = Role.Button
            },
        ) {
            Text(
                text = label,
                style = TextStyle(
                    fontSize = CivTokens.Typography.FontSize.base,
                    fontWeight = FontWeight.SemiBold,
                ),
            )
        }
    } else {
        OutlinedButton(
            onClick = onClick,
            enabled = enabled,
            border = BorderStroke(CivTokens.Border.Width._2, primaryColor),
            colors = ButtonDefaults.outlinedButtonColors(
                contentColor = primaryColor,
            ),
            modifier = Modifier.semantics {
                selected = false
                role = Role.Button
            },
        ) {
            Text(
                text = label,
                style = TextStyle(
                    fontSize = CivTokens.Typography.FontSize.base,
                    fontWeight = FontWeight.SemiBold,
                ),
            )
        }
    }
}

// MARK: - Preview

@Preview(showBackground = true, name = "CivYesNo")
@Composable
private fun CivYesNoPreview() {
    Column(modifier = Modifier.padding(16.dp)) {
        var citizen by remember { mutableStateOf("") }
        CivYesNo(
            legend = "Are you a U.S. citizen?",
            value = citizen,
            onValueChange = { citizen = it },
            hint = "Select one option",
            required = true,
        )

        var veteran by remember { mutableStateOf("yes") }
        CivYesNo(
            legend = "Are you a veteran?",
            value = veteran,
            onValueChange = { veteran = it },
        )

        var disability by remember { mutableStateOf("") }
        CivYesNo(
            legend = "Do you have a service-connected disability?",
            value = disability,
            onValueChange = { disability = it },
            unsureLabel = "I'm not sure",
        )

        CivYesNo(
            legend = "Disabled question",
            value = "",
            onValueChange = {},
            disabled = true,
        )

        CivYesNo(
            legend = "With error",
            value = "",
            onValueChange = {},
            error = "Please select an option",
            required = true,
        )
    }
}
